### PR TITLE
Text.format() feature

### PR DESCRIPTION
**Pull request by pom11**
## Type of changes

- [ ] Bug fix
- [x] New feature
- [ ] Documentation / docstrings
- [X] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description
Added `format` method for `Text`. Currently works just for string placeholders like `{name}`. 

Supported placeholders: `{string}`
Unsupported placeholders: `{0}` `{}`

TODO:
Support all types of placeholders

### CODE REVIEW

**Comment by willmcgugan** [comment]
I'm afraid it doesn't work as expected.

Screenshot 2023-03-27 at 13 56 06

It's a worthwhile feature, but it would require full understanding of the Text internals. We would also need comprehensive unit tests.

**Comment by pom11** [comment]
will take a look asap

**Comment by pom11** [comment]
Screenshot 2023-03-27 at 16 14 11
Screenshot 2023-03-27 at 16 14 42
seems to be working for my machine

**Comment by pom11** [comment]
Screenshot 2023-03-27 at 16 46 28
found the problem. 
```
>>> t
``` messes everything up

**Comment by pom11** [comment]
solved.

File: rich/text.py
Comment (willmcgugan) [inline rich/text.py:450]: Not clear what is going on here.

File: tests/test_text.py
Comment (willmcgugan) [inline tests/test_text.py:126]: This needs more tests.

Check plain strings, strings with a single substitution, no styles, several styles, overlapping styles, etc. Try to cover as many edge cases as you can.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
The efficiency of this method could be improved.

Readability could use some work. Better named variables etc. `str_index` says what it is, but doesn't say what it is used for. We have a no abbreviation and no single letter variables (with a few exceptions).

A few  well-placed comments would also help to follow the code.

We will need a lot more tests before this could be merged.

File: rich/text.py
Comment (pom11) [reply rich/text.py:469]: yes

File: rich/text.py
Comment (pom11) [reply rich/text.py:450]: converted Span to a list cause can't set attributes to them

**Comment by pom11** [comment]
will refactor as discussed on discord

**Comment by pom11** [comment]
LGTM

**Comment by pom11** [comment]
updates?

**Comment by willmcgugan** [comment]
Sorry, I'm afraid I can't merge this PR. I have a feeling the implementation is doing more work than it needs to be, and the overall code quality isn't quite high enough.

The tests don't give me any confidence in the code. Test should be treated almost like documentation, with a function for each aspect of the code being tested. It's not enough just to get full converage.

If you need this, you might want to factor it out in to your own code. Keep it as an external function rather than a method in the core lib.

**Comment by pom11** [comment]
the code is pretty much straight forward, covering all aspects from [docs](https://docs.python.org/3/library/string.html#format-string-syntax). at first glance the complexity comes from treating mixed arguments such as Text with string. i admit the missing comments from test could mislead what feature is testing.

do you plan to add this feature in the nearby future? would love to see a better approach

tried my best.

**Pull request closed without merging**
